pkg/cmd/cli/cmd/extract.go

```diff
@@ -0,0 +1,171 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	extractLong = `
+Extract files out of secrets and config maps
+
+The extract command makes it easy to download the contents of a config map or secret into a directory.
+Each key in the config map or secret are created as a separate file with the name of the key, as they
+are when you mount a secret or config map into a container.
+
+You can limit which keys are extracted with the --only=NAME flag, or set the directory to extract to
+with --to=DIRECTORY.`
+
+	extractExample = `  # extract the secret "test" to the current directory
+  %[1]s extract secret/test
+
+  # extract the config map "nginx" to the /tmp directory
+  %[1]s extract configmap/nginx --to=/tmp`
+)
+
+type ExtractOptions struct {
+	Out, Err        io.Writer
+	Filenames       []string
+	OnlyValues      []string
+	TargetDirectory string
+	Overwrite       bool
+
+	VisitorFn             func(resource.VisitorFunc) error
+	ExtractFileContentsFn func(runtime.Object) (map[string][]byte, bool, error)
+}
+
+func NewCmdExtract(fullName string, f *clientcmd.Factory, in io.Reader, out, errOut io.Writer) *cobra.Command {
+	options := &ExtractOptions{
+		Out: out,
+		Err: errOut,
+	}
+	cmd := &cobra.Command{
+		Use:     "extract RESOURCE/NAME [--to=DIRECTORY] [options]",
```

Comment:
You don't need `[options]`, Cobra adds it to the end if it's not in the string yet.